grails-app/services/com/unifina/service/StreamService.groovy

```diff
@@ -140,23 +140,25 @@ class StreamService {
 		}
 
 		if (key != null) {
-			if (isDirectPermissionToStream(key, stream)) {
+			if (permissionService.check(key, stream, Permission.Operation.STREAM_GET)) {
 				action.call(stream)
 			} else {
-				throw new NotPermittedException(null, "Stream", id, Permission.Operation.READ.id)
+				throw new NotPermittedException(null, "Stream", id, Permission.Operation.STREAM_GET.id)
 			}
 		} else {
-			if (isDirectPermissionToStream(user, stream) || isPermissionToStreamsUiChannelCanvas(user, stream) || isPermissionToStreamViaDashboard(user, stream)) {
+			boolean hasPermissionToStream = permissionService.check(user, stream, Permission.Operation.STREAM_GET)
+			boolean hasPermissionToStreamsUiChannelCanvas = stream.uiChannel && stream.uiChannelCanvas && permissionService.check(user, stream.uiChannelCanvas, Permission.Operation.CANVAS_GET)
```

Comment:
This is very minor, but I noticed that in the old code, the UI channel check doesn't get executed if the user has direct permission. This saves us one call to `permissionService.check(...)`, which is expensive.

You could get the same optimization by changing the second line to
```
boolean hasPermissionToStreamsUiChannelCanvas = !hasPermissionToStream && stream.uiChannel && stream.uiChannelCanvas && permissionService.check(user, stream.uiChannelCanvas, Permission.Operation.CANVAS_GET)
```

BUT I'm not entirely sure if the new code is actually more readable than the old one? (I think it was a wise choice to delete `isDirectPermissionToStream` though)

Reply:
@hpihkala I'll optimise it performance. Readability doesn't really suffer that much.